src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs

```diff
@@ -249,16 +237,8 @@ public void TestPreviewDiagnostic()
                         var rightTagger = rightProvider.CreateTagger<IErrorTag>(rightBuffer);
                         using (var rightDisposable = rightTagger as IDisposable)
                         {
-                            // wait up to 20 seconds for diagnostics
-                            taskSource.Task.Wait(20000);
-                            if (!taskSource.Task.IsCompleted)
```

Comment:
is this okay? it was waiting for analysis to actually happen. I think your approach seems good. just that TestWorkspace the test creates, you need to set the workspace kind as preview. and it needs to use its own ExportProvider otherwise, listener will get messed up since it got shared with all other tests. and nit, but use FeatureAttributes.SolutionCrawler rather than DiagnosticService.

Reply:
➡️ It's waiting for `DiagnosticService` because that's the feature reported by `PreviewSolutionCrawlerRegistrationService`.

Reply:
yea. I got it. the CreateChangedDocumentPreviewViewAsync internally creates the PreviewWorkspace.